File: javascript/beginner/palindrome-linked-list/palindrome-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if (head === null || head.next === null) {
        return true;
    }

    // 1. Find the middle of the list
    let slow = head;
    let fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
    }

    // 2. Reverse the second half
    let prev = null;
    let current = slow;
    while (current !== null) {
        const nextNode = current.next;
        current.next = prev;
        prev = current;
        current = nextNode;
    }
    // `prev` is now the head of the reversed second half

    // 3. Compare the first and reversed second halves
    let firstHalfPtr = head;
    let secondHalfPtr = prev;
    while (secondHalfPtr !== null) {
        if (firstHalfPtr.val !== secondHalfPtr.val) {
            return false;
        }
        firstHalfPtr = firstHalfPtr.next;
        secondHalfPtr = secondHalfPtr.next;
    }

    return true;
};

module.exports = isPalindrome;

File: python/beginner/palindrome-linked-list/palindrome_linked_list.py
"""
Palindrome Linked List

Given the head of a singly linked list, return true if it is a palindrome 
or false otherwise.
"""

import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), '../../..'))
from utils import ListNode


def isPalindrome(head):
    """
    Check if a linked list is a palindrome.
    
    Args:
        head: Head of the linked list
        
    Returns:
        True if the linked list is a palindrome, False otherwise
        
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not head or not head.next:
        return True
    
    # 1. Find the middle of the list using two pointers
    slow = head
    fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    
    # 2. Reverse the second half
    prev = None
    current = slow
    while current:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node
    # prev is now the head of the reversed second half
    
    # 3. Compare the first and reversed second halves
    first_half_ptr = head
    second_half_ptr = prev
    while second_half_ptr:
        if first_half_ptr.val != second_half_ptr.val:
            return False
        first_half_ptr = first_half_ptr.next
        second_half_ptr = second_half_ptr.next
    
    return True